### PR TITLE
API: Euclidean Geometry Updates

### PR DESCRIPTION
**Pull request by nyoungbq**
- Removed Point2D and Point3D classes -> Now a templated type-alias for `Vec3`
- Updated BoundingBox class to interface with Vec2 and Vec3 and standardized usage throughout code base
- Ported and updated a majority of the GeometryMath class from SIMPL
- Extracted non-templated function definitions out of GeometryUtilities header into a cpp file

## Naming Conventions

Naming of variables should descriptive where needed. Loop Control Variables can use `i` if warranted. Most of these conventions are enforced through the clang-tidy and clang-format configuration files. See the file `complex/docs/Code_Style_Guide.md` for a more in depth explanation.

## Filter Checklist

The help file `complex/docs/Porting_Filters.md` has documentation to help you port or write new filters. At the top is a nice checklist of items that should be noted when porting a filter.

## Unit Testing

The idea of unit testing is to test the filter for proper execution and error handling. How many variations on a unit test each filter needs is entirely dependent on what the filter is doing. Generally, the variations can fall into a few categories:

- [x] 1 Unit test to test output from the filter against known exemplar set of data
- [x] 1 Unit test to test invalid input code paths that are specific to a filter. Don't test that a DataPath does not exist since that test is already performed as part of the SelectDataArrayAction.

## Code Cleanup
- [x] No commented out code (rare exceptions to this is allowed..)
- [x] No API changes were made (or the changes have been approved)
- [x] No major design changes were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added license to new files (if any)
- [x] Added example pipelines that use the filter
- [x] Classes and methods are properly documented

### CODE REVIEW

File: src/complex/Common/BoundingBox.hpp [hunk 1]
```diff
@@ -2,20 +2,22 @@
...
 
-#include "complex/Common/Point3D.hpp"
+#include "complex/Common/Array.hpp"
 
 namespace complex
 {
-
 /**
  * @class BoundingBox
  * @brief The BoundingBox class is designed to describe a box in 3D space
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]: Since this supports 2D now, the documentation should be updated.

File: src/complex/Common/BoundingBox.hpp [hunk 2]
```diff
@@ -49,29 +47,45 @@ class BoundingBox
...
 
   /**
-   * @brief Constructs a new BoundingBox defined by the array of position values.
+   * @brief Constructs a new BoundingBox3D defined by the array of position values.
    * The format is min X, min Y, min Z, max X, max Y, max Z.
    * @param arr
    */
-  BoundingBox(const std::array<T, 6>& arr)
-  : m_Bounds(arr)
+  explicit BoundingBox(const std::array<T, 6>& arr)
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]:
```suggestion
  template <class = std::enable_if_t<PointType::Dimensions == 6>>
  explicit BoundingBox(const std::array<T, 6>& arr)
```
We should conditionally enable these constructors depending on the point type. This will require a static constexpr member in the `Vec` class.

File: src/complex/Common/BoundingBox.hpp [hunk 3]
```diff
@@ -49,29 +47,45 @@ class BoundingBox
...
+  , m_Upper(Point3D<T>(arr[3], arr[4], arr[5]))
+  {
+  }
+
+  /**
+   * @brief Constructs a new BoundingBox2D defined by the array of position values.
+   * The format is min X, min Y, max X, max Y.
+   * @param arr
+   */
+  explicit BoundingBox(const std::array<T, 4>& arr)
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]:
```suggestion
  template <class = std::enable_if_t<PointType::Dimensions == 4>>
  explicit BoundingBox(const std::array<T, 4>& arr)
```
See previous.

File: src/complex/Common/BoundingBox.hpp [hunk 4]
```diff
@@ -49,29 +47,45 @@ class BoundingBox
...
-   * std::array<T, 6> constructor, min X, min Y, min Z, max X, max Y, max Z.
+   * The pointer is assumed to contain either 4 (2D) or 6 (3D) values formatted the same as the
+   * std::array<T, 6> constructor, min X, min Y, max X, max Y or the
+   * std::array<T, 6> constructor, min X, min Y, min Z, max X, max Y, max Z
+   * respectively
    * @param arr
    */
-  BoundingBox(Pointer arr)
-  : m_Bounds({arr[0], arr[1], arr[2], arr[3], arr[4], arr[5]})
+  explicit BoundingBox(Pointer arr)
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]:
```suggestion
  explicit BoundingBox(const Pointer arr)
```

File: src/complex/Common/BoundingBox.hpp [hunk 5]
```diff
@@ -81,145 +95,62 @@ class BoundingBox
    * @brief Returns the Min point
    * @return
    */
-  std::array<ValueType, 3> getMinPoint() const
+  PointType getMinPoint() const
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]:
```suggestion
  const PointType& getMinPoint() const
```
Since we're changing the class anyways, I'd make the getters return by const ref.

File: src/complex/Common/BoundingBox.hpp [hunk 6]
```diff
@@ -81,145 +95,62 @@ class BoundingBox
...
-    return {m_Bounds[0], m_Bounds[1], m_Bounds[2]};
+    return m_Lower;
   }
 
   /**
    * @brief Returns the Max point
    * @return
    */
-  std::array<ValueType, 3> getMaxPoint() const
+  PointType getMaxPoint() const
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]:
```suggestion
  const PointType& getMaxPoint() const
```
See previous.

File: src/complex/Common/BoundingBox.hpp [hunk 7]
```diff
@@ -81,145 +95,62 @@ class BoundingBox
...
+   * @return
    */
-  void setMaxZ(ValueType value)
+  PointType center() const
   {
-    m_Bounds[5] = value;
+    PointType boxSize = sideLengths();
+    if constexpr(std::is_same_v<Point<ValueType>, Point3D<ValueType>>)
+    {
+      return {m_Lower[0] + boxSize[0] / 2.0, m_Lower[1] + boxSize[1] / 2.0, m_Lower[2] + boxSize[2] / 2.0};
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]: Maybe we should have a division operator on `Vec`. I see we have multiplication already.

File: src/complex/Common/BoundingBox.hpp [hunk 8]
```diff
@@ -269,7 +175,8 @@ class BoundingBox
    */
   BoundingBox& operator=(const BoundingBox& rhs)
   {
-    m_Bounds = rhs.m_Bounds;
+    m_Lower = rhs.m_Lower;
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]: I think we can default all the copy/move constructors/assignment.

File: src/complex/Utilities/Math/GeometryMath.hpp [hunk 9]
```diff
@@ -23,12 +24,18 @@ namespace GeometryMath
...
- * @return float
+ * @return T
  */
 template <typename T>
-T CosThetaBetweenVectors(const complex::Point3D<T>& a, const complex::Point3D<T>& b)
+inline T CosThetaBetweenVectors(const complex::Point3D<T>& a, const complex::Point3D<T>& b)
 {
-  throw std::runtime_error("");
+  T norm1 = sqrt(a.sumOfSquares());
+  T norm2 = sqrt(b.sumOfSquares());
```
Comment (JDuffeyBQ) [inline src/complex/Utilities/Math/GeometryMath.hpp]:
```suggestion
  T norm1 = std::sqrt(a.sumOfSquares());
  T norm2 = std::sqrt(b.sumOfSquares());
```
All `<cmath>` functions should have the `std::` namespace prefixed.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

File: src/complex/Common/BoundingBox.hpp [hunk 10]
```diff
@@ -49,30 +47,56 @@ class BoundingBox
...
+   * std::array<T, 4> constructor, min X, min Y, max X, max Y or the
+   * std::array<T, 6> constructor, min X, min Y, min Z, max X, max Y, max Z
+   * respectively
    * @param arr
    */
-  BoundingBox(Pointer arr)
-  : m_Bounds({arr[0], arr[1], arr[2], arr[3], arr[4], arr[5]})
+  explicit BoundingBox(const Pointer arr)
   {
+    if(arr->size() == 6)
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]: `Pointer` is just `T` right?

File: src/complex/Common/Array.hpp [hunk 11]
```diff
@@ -439,6 +439,118 @@ class Vec2 : public Array<T, 2>
...
+  inline Vec2 operator*(const T r) const
+  {
+    return Vec2((*this)[0] * r, (*this)[1] * r);
+  }
+  /**
+   * @brief Multiples this Vec2 with another Vec2 using element wise multiplication and returns a new instance
+   * @param Vec2
+   * @return new Vec2 instance with the result
+   */
+  inline Vec2 operator*(const Vec2& v) const
```
Comment (JDuffeyBQ) [inline src/complex/Common/Array.hpp]: I know `operator*` with `Vec3` was added previously, but I'm thinking it might not be obvious that it's element-wise multiplication. That might be a topic for another PR though.

File: src/complex/Common/BoundingBox.hpp (hunk 10 above)
Comment (nyoungbq) [reply src/complex/Common/BoundingBox.hpp]: @JDuffeyBQ `Pointer` is `T*` || `ValueType` is `T`

File: src/complex/Common/Array.hpp (hunk 11 above)
Comment (nyoungbq) [reply src/complex/Common/Array.hpp]: I don't mind refactoring it to a different operator in this PR what would you suggest the operator should be used to make this operation more clear? Or should I just make this a function such as `elemMulOp()`

Comment (JDuffeyBQ) [reply src/complex/Common/Array.hpp]: I'd say just a normal function.

File: src/complex/Common/BoundingBox.hpp (hunk 10 above)
Comment (JDuffeyBQ) [reply src/complex/Common/BoundingBox.hpp]: Then shouldn't this constructor take a `span` instead if we want to check the size at runtime?

Comment (nyoungbq) [reply src/complex/Common/BoundingBox.hpp]: Yes I will fix that

File: src/complex/Common/BoundingBox.hpp [hunk 12]
```diff
@@ -49,30 +49,56 @@ class BoundingBox
...
-   * std::array<T, 6> constructor, min X, min Y, min Z, max X, max Y, max Z.
+   * The pointer is assumed to contain either 4 (2D) or 6 (3D) values formatted the same as the
+   * std::array<T, 4> constructor, min X, min Y, max X, max Y or the
+   * std::array<T, 6> constructor, min X, min Y, min Z, max X, max Y, max Z
+   * respectively
    * @param arr
    */
-  BoundingBox(Pointer arr)
-  : m_Bounds({arr[0], arr[1], arr[2], arr[3], arr[4], arr[5]})
+  explicit BoundingBox(nonstd::span<ValueType> arr)
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]:
```suggestion
  explicit BoundingBox(nonstd::span<const ValueType> arr)
```

File: src/complex/Common/BoundingBox.hpp [hunk 13]
```diff
@@ -49,30 +49,56 @@ class BoundingBox
...
   {
+    if(arr->size() == 6)
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]:
```suggestion
    if(arr.size() == 6)
```

File: src/complex/Common/BoundingBox.hpp [hunk 14]
```diff
@@ -49,30 +49,56 @@ class BoundingBox
...
+    {
+      m_Lower = Point3D<T>(arr[0], arr[1], arr[2]);
+      m_Upper = Point3D<T>(arr[3], arr[4], arr[5]);
+    }
+    else if(arr->size() == 4)
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]:
```suggestion
    else if(arr.size() == 4)
```

File: src/complex/Common/BoundingBox.hpp [hunk 15]
```diff
@@ -49,30 +49,56 @@ class BoundingBox
...
+    {
+      m_Lower = Point3D<T>(arr[0], arr[1]);
+      m_Upper = Point3D<T>(arr[2], arr[3]);
```
Comment (JDuffeyBQ) [inline src/complex/Common/BoundingBox.hpp]: Actually looking at this again, shouldn't these in this branch be `Point2D`? And if so, that means we can't branch at runtime since there are different point types.

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Review by JDuffeyBQ — APPROVED** [review]
